File: jvm/selfie-lib/src/commonMain/kotlin/com/diffplug/selfie/guts/EscapeLeadingWhitespace.kt
/*
 * Copyright (C) 2024 DiffPlug
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.diffplug.selfie.guts

internal enum class EscapeLeadingWhitespace {
  ALWAYS,
  NEVER,
  ONLY_ON_SPACE,
  ONLY_ON_TAB;
  fun escapeLine(line: String, space: String, tab: String): String =
      if (line.startsWith(" ")) {
        if (this == ALWAYS || this == ONLY_ON_SPACE) "$space${line.drop(1)}" else line
      } else if (line.startsWith("\t")) {
        if (this == ALWAYS || this == ONLY_ON_TAB) "$tab${line.drop(1)}" else line
      } else line

  companion object {
    private val MIXED = 'm'
    fun appropriateFor(fileContent: String): EscapeLeadingWhitespace {
      val commonWhitespace =
          fileContent
              .lineSequence()
              .mapNotNull { line ->
                val whitespace = line.takeWhile { it.isWhitespace() }
                if (whitespace.isEmpty() || whitespace == " ") null
                else if (whitespace.all { it == ' ' }) ' '
                else if (whitespace.all { it == '\t' }) '\t' else MIXED
              }
              .reduceOrNull { a, b -> if (a == b) a else MIXED }
      return if (commonWhitespace == ' ') ONLY_ON_TAB
      else if (commonWhitespace == '\t') ONLY_ON_SPACE else ALWAYS
    }
  }
}

File: python/selfie-lib/selfie_lib/EscapeLeadingWhitespace.py
from enum import Enum, auto


class EscapeLeadingWhitespace(Enum):
    ALWAYS = auto()
    NEVER = auto()
    ONLY_ON_SPACE = auto()
    ONLY_ON_TAB = auto()

    def escape_line(self, line: str, space: str, tab: str) -> str:
        if line.startswith(" "):
            if (
                self == EscapeLeadingWhitespace.ALWAYS
                or self == EscapeLeadingWhitespace.ONLY_ON_SPACE
            ):
                return f"{space}{line[1:]}"
            else:
                return line
        elif line.startswith("\t"):
            if (
                self == EscapeLeadingWhitespace.ALWAYS
                or self == EscapeLeadingWhitespace.ONLY_ON_TAB
            ):
                return f"{tab}{line[1:]}"
            else:
                return line
        else:
            return line

    @classmethod
    def appropriate_for(cls, file_content: str) -> "EscapeLeadingWhitespace":
        MIXED = "m"
        common_whitespace = None

        for line in file_content.splitlines():
            whitespace = "".join(c for c in line if c.isspace())
            if not whitespace:
                continue
            elif all(c == " " for c in whitespace):
                whitespace = " "
            elif all(c == "\t" for c in whitespace):
                whitespace = "\t"
            else:
                whitespace = MIXED

            if common_whitespace is None:
                common_whitespace = whitespace
            elif common_whitespace != whitespace:
                common_whitespace = MIXED
                break

        if common_whitespace == " ":
            return cls.ONLY_ON_TAB
        elif common_whitespace == "\t":
            return cls.ONLY_ON_SPACE
        else:
            return cls.ALWAYS
